torch/_functorch/aot_autograd.py

```diff
@@ -4456,9 +4462,9 @@ def convert(idx, x):
 
         fake_flat_args = process_inputs(flat_args)
 
-        needs_autograd = (
+        needs_autograd = aot_config.allow_joint_graph and (
```

Comment:
Hey sorry for dropping this @jon-chuang.

Hmm, I don't love the idea of a config - ideally we should just detect whether we need a training vs. inference graph in all cases automatically.

Now that we are basically detecting "do we need a training graph" in two steps, the first step here is basically just a heuristic (if we do need a training graph then this should definitely be set to True, but if it turn out that we don't need a training graph, because all the graph outputs don't require grad, then it's ok to pessimistically set this to true, and set it to false lower down).

So in theory we could just set this to `True` here, although one downside right now is that getting it wrong means that we need to re-run `run_functionalized_fw_and_collect_metadata` in the subclass path, which will hurt compile perf.

Reply:
Instead of the new config, can you just make the condition this?
```
needs_autograd = any(x.requires_grad for x in fake_flat_args if isinstance(x, Tensor))
# forget about looking at `is_grad_enabled`
```

That should fix this case too (I had the same idea [here](https://github.com/pytorch/pytorch/pull/115294), but happy to land you PR), since we won't incorrect label the test you've written as an inference graph.